File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

// AM: 15       34/37
// VS: 13/16    4
// MC: 16       51/52/53/54/55

int main(void)
{
    long long int number = get_long_long("Number: ");
    int cardlength = 0;
    int sum = 0;
    int digit = 0;
    int second = 0;
    int even = 0;

    while (number > 0)
    {
        cardlength++;
        // variable for AM/MC
        second = digit;
        // start from last
        digit = number % 10;
        number = (number - digit) / 10;

        // each even digit
        if (cardlength % 2 == 0)
        {
            even = digit * 2;
            if (even >= 10)
            {
                // from 2 * 5 to 2 * 9
                sum += even % 10;
                sum += 1;
            }
            else
            {
                sum += even;
            }
        }
        else
        {
            sum += digit;
        }
    }
    // digit == the start number
    // initial the answer
    string result = "INVALID";
    // do the start number check with sum
    if (sum % 10 == 0)
    {
        // VISA 13/16 digits with 4
        if ((cardlength == 13 || cardlength == 16) && digit == 4)
        {
            result = "VISA";
        }
        // AMEX 15 digits with 34/37
        else if (cardlength == 15 && digit == 3 && (second == 4 || second == 7))
        {
            result = "AMEX";
        }
        // MC 16 digits with 51-55
        else if (cardlength == 16 && digit == 5 && (second >= 1 && second <= 5))
        {
            result = "MASTERCARD";
        }
    }

    printf("%s\n", result);
}

File: pset6/credit/credit.py
from cs50 import get_string
from math import floor

# AM: 15       34/37
# VS: 13/16    4
# MC: 16       51/52/53/54/55

result = "INVALID"
# get input
card_number = get_string("Number: ")
card_number = card_number.strip()
card_length = len(card_number)

# length == 15
if (card_length == 15):
    if (card_number[:2] == "34" or card_number[:2] == "37"):
        result = "AMEX"
# length == 13
elif (card_length == 13):
    if (card_number[:1] == "4"):
        result = "VISA"
# length == 16
elif (card_length == 16):
    # VISA
    if (card_number[:1] == "4"):
        result = "VISA"
    # MC
    elif (card_number[:2] == "51" or card_number[:2] == "52" or card_number[:2] == "53" or card_number[:2] == "54" or card_number[:2] == "55"):
        result = "MASTERCARD"

#############################
# do the digits calculation #
#############################
if (result != "INVALID"):
    card_number = int(card_number)
    last = 0
    temp = 0
    seco = 0

    while (card_number > 0):
        last += card_number % 10  # last digit
        card_number /= 10
        card_number = floor(card_number)
        temp = (card_number % 10) * 2  # do the digit multiple
        # result >= 10
        if (temp >= 10):
            seco += temp % 10
            seco += 1
        else:
            seco += temp
        card_number /= 10
        card_number = floor(card_number)
    # valid card
    if ((last + seco) % 10 == 0):
        print(result)
    else:
        print("INVALID")
else:
    print(result)